src/v/cluster/data_migration_backend.h

```diff
@@ -89,11 +90,22 @@ class backend {
         errc ec;
     };
     class topic_scoped_work_state {
+        using self = topic_scoped_work_state;
         ss::abort_source _as;
+        retry_chain_node _rcn;
+        ss::shared_promise<errc> _promise;
```

Comment:
maybe we should consider using a gate here ? Also why shared promise is required ? The shared promise has one caveat that it does not immediatelly return `true` from `available()` method after the `set()` is called

Reply:
This entity works as a guard for no more than one fiber to own an map entry and to perform work. Mutliple fibres can compete to have/make an entry if the coordinator quickly realizes that a migration has went forward, presumably by another coordinator when leadership changes. E.g.
1) we start working towards prepared state
2) an update has come that the migration is in executing state, so we need to cancel existing work and spawn new work towards executed state
3) yet another update has come that the migration has finished, so we need to stop the work
(2) and (3) will be both waiting for (1) to complete.

I thought of using mutexes or gates + a member for storing the last result. I felt like reinventing shared future/promise. We do not check the future availability where we set its value.

Reply:
Slightly tangential
> The shared promise has one caveat that it does not immediatelly return true from available() method after the set() is called

This is no longer an issue after recent Seastar rebase that pulled in https://github.com/scylladb/seastar/commit/ce96371f604803d4b2816479bfcd2a59d1f8d767